### PR TITLE
delete intermediate derivatives for memory savings

### PR DESCRIPTION
**Pull request by ryanwebster90**
adds a conserveMemory property to Net, similar to MatConvNet dagnn, specifically for the
deletion of intermediate derivatives during back prop.

### CODE REVIEW

File: matlab/@Net/eval.m [hunk 1]
```diff
@@ -164,6 +165,19 @@ function eval(net, inputs, mode, derOutput, accumulateParamDers)
           vars{inputDer} = vars{inputDer} + slice_der(args{:}) ;
         end
       end
+      
+      % remove derivatives (even input vars)
+      % as they are no longer used in the network
+    if conserveMemory
+        for i = numel(layer.inputVars):-1:1
+            if ~mod(layer.inputVars(i),2)
+                vars{layer.inputVars(end)} = [];
```
Comment (jotaf98) [inline matlab/@Net/eval.m]: Should this "end" maybe be "i"?

File: matlab/@Net/eval.m [hunk 2]
```diff
@@ -164,6 +165,19 @@ function eval(net, inputs, mode, derOutput, accumulateParamDers)
           vars{inputDer} = vars{inputDer} + slice_der(args{:}) ;
         end
       end
+      
+      % remove derivatives (even input vars)
+      % as they are no longer used in the network
+    if conserveMemory
+        for i = numel(layer.inputVars):-1:1
```
Comment (jotaf98) [inline matlab/@Net/eval.m]: Maybe this loop can be vectorized? Suggestion (assuming I understand the logic correctly) :
```
idx = (mod(layer.inputVars, 2) == 0) ;  % logical vector with 1 where inputVars must be deleted
vars(layer.inputVars(idx)) = [] ;  % delete all at once
```

**Review by jotaf98 — COMMENTED** [review]
Good addition! I made a few suggestions for improvement, not absolutely necessary but nice-to-have ;)

**Comment by jotaf98** [comment]
Link to in-depth discussion:
https://github.com/vlfeat/autonn/issues/12

File: matlab/@Net/eval.m (hunk 2 above)
Comment (ryanwebster90) [reply matlab/@Net/eval.m]: Sure - it's now vectored - but using the syntax
` vars(layer.inputVars(idx)) = {[]};`
other wise MATLAB deletes the cell entry.

File: matlab/@Net/eval.m (hunk 1 above)
Comment (ryanwebster90) [reply matlab/@Net/eval.m]: Yep - it's fixed

**Pull request merged**